lib/algolia.ex

```diff
@@ -204,22 +206,21 @@ defmodule Algolia do
   end
 
   defp do_send_request(subdomain_hint, request, curr_retry) do
-    url = request_url(subdomain_hint, curr_retry, request[:path])
-    headers = request_headers(request[:options] || [])
-    body = request[:body] || ""
-
-    request[:method]
-    |> :hackney.request(url, headers, body, [
-      :with_body,
-      path_encode_fun: &URI.encode/1,
-      connect_timeout: 3_000 * (curr_retry + 1),
-      recv_timeout: 30_000 * (curr_retry + 1)
-    ])
+    {path, req} = Map.pop(request, :path)
+    url = request_url(subdomain_hint, curr_retry, path)
+
+    {options, req} = Map.pop(req, :options, [])
+    headers = request_headers(options)
+
+    req
+    |> Map.to_list()
+    |> Enum.concat(url: url, headers: headers)
+    |> request()
```

Comment:
My goal was to leave the API the same as before, and the existing API wouldn't be able to accommodate passing a client around. Maybe in a future change, we can add an optional client parameter to all the functions, so that middleware can be customized.

I do plan to introduce some internal middleware using the `plug` macro, but that doesn't require passing clients around either.